src/components/Toolbar.js

```diff
@@ -11,13 +36,82 @@ const shortcutIds = [
   'search',
 ]
 
-export const Toolbar = connect(({ settings: { dark } }) => ({ dark }))(({ dark }) =>
-  <div className="toolbar">
-    {shortcutIds.map(id => {
-      const { name, svg: Icon, exec } = shortcutById(id)
-      return <div key={id} className="toolbar-icon" title={name} onClick={exec}>
-        <Icon fill={dark ? 'white' : 'black'} />
+const ToolbarMessageOverlay = ({
+  classname = 'toolbar-overlay',
+  id
+}) => {
+  const { name, description } = shortcutById(id)
+  return (
+    <div className={classname}>
+      <div className={'overlay-name'}>{name}</div>
+      <div className={'overlay-body'}>{description}</div>
+    </div>
+  )
+}
+
+let holdTimer
+
+const onHoldDownShortcut = id => {
+  // on chrome setTimeout doesn't seem to work on the first click, clearing it before hand fixes the problem
+  clearTimeout(holdTimer)
+  holdTimer = setTimeout(() => {
+    overlayReveal(id)
+  }, SHORTCUT_HINT_OVERLAY_TIMEOUT)
+}
+
+export const Toolbar = connect(({ toolbarOverlay, settings: { dark } }) => ({
+  dark,
+  toolbarOverlay
+}))(
+  ({
+    dark,
+    toolbarOverlay: {
+      shortcutId
+    }
+  }) => (
+    <div>
+      <div className={'toolbar-container'}>
+        <div className={'toolbar'}>
+          {shortcutIds.map(id => {
+            const { name, svg: Icon, exec } = shortcutById(id)
+            return (
+              <div
+                key={name}
+                id={id}
+                className='toolbar-icon'
+                onMouseDown={() => onHoldDownShortcut(id)}
+                onMouseOver={() => {
+                    if (shortcutId) overlayReveal(id)
+                  }
+                }
+                onTouchStart={() => onHoldDownShortcut(id)}
+                onClick={() => exec(id)}
+              >
+                <Icon id={id} fill={dark ? 'white' : 'black'} />
+              </div>
+            )
+          })}
+        </div>
+        <TransitionGroup>
+          <CSSTransition key={0} timeout={200} classNames='fade'>
+            {!isTouchEnabled() && shortcutId ? (
+              <ToolbarMessageOverlay
+                id={shortcutId}
+              />
+            ) : <span></span>}
+          </CSSTransition>
+        </TransitionGroup>
       </div>
-    })}
-  </div>
+      <TransitionGroup>
+        <CSSTransition key={0} timeout={200} classNames='fade'>
+          {isTouchEnabled() && shortcutId ? (
+            <ToolbarMessageOverlay
+              classname={'touch-toolbar-overlay'}
+              id={shortcutId}
+            />
+          ) : <span></span>}
+        </CSSTransition>
+      </TransitionGroup>
```

Comment:
These blocks of code are completely identical, so we'll want to abstract it into a local component.

Reply:
Got it.

Reply:
<img width="716" alt="Screen Shot 2020-01-08 at 6 10 22 PM" src="https://user-images.githubusercontent.com/12413798/71989149-2f3d5380-3242-11ea-8ee4-9a47dbe56b66.png">
Does this work for you?